### PR TITLE
Add unit test to make sure helloworld samples README consistent with source

### PR DESCRIPTION
**Pull request by chaodaiG**
In PR https://github.com/knative/docs/pull/1092, there were a few inconsistencies discovered between README files and source files. Add unit test to checks contents of README.md files, and ensures that the real code of the samples is properly embedded in the README.md files

### CODE REVIEW

File: test/sampleconsistency/sampleconsistency_test.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+func normalize(si string) string {
+	return strings.Trim(si, " `\t\r\n")
+}
+
+func checkContains(t *testing.T, rl []string, src string) {
+	sl := readlines(t, src)
+	ir := 0
+	is := 0
+	best := -1
+	for ir < len(rl) && is < len(sl) {
```
Comment (adrcunha) [inline test/sampleconsistency/sampleconsistency_test.go]: This loop would benefit from a comment.

File: test/sampleconsistency/sampleconsistency_test.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+		} else {
+			is++
+			if is > best {
+				best = is
+			}
+		}
+		ir++
+	}
+	if is < len(sl) && best < len(sl) && best != -1 {
+		t.Fatalf("README missing line '%s' '%d' in file '%s'", sl[best], best, src)
```
Comment (adrcunha) [inline test/sampleconsistency/sampleconsistency_test.go]:
```suggestion
		t.Fatalf("README.md file is missing line %d ('%s') from file '%s'", best, sl[best], src)
```

File: test/sampleconsistency/sampleconsistency_test.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+func checkDoc(t *testing.T, lc sampleapp.LanguageConfig) {
+	readme := path.Join(lc.SrcDir, "README.md")
+	rl := readlines(t, readme)
+	for _, f := range lc.Copies {
+		src := path.Join(lc.SrcDir, f)
+		checkContains(t, rl, src)
+	}
+}
+
+// TestDocSrc runs all sample apps from different languages
```
Comment (adrcunha) [inline test/sampleconsistency/sampleconsistency_test.go]:
```suggestion
// TestDocSrc checks the contents of sample apps in the respective README.md files
```

File: test/sampleconsistency/sampleconsistency_test.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+func TestDocSrc(t *testing.T) {
+	lcs, err := sampleapp.GetConfigs(configFile)
+	if nil != err {
+		t.Fatalf("Failed reading config file %s: '%v'", configFile, err)
+	}
+
+	whitelist := make(map[string]bool)
```
Comment (adrcunha) [inline test/sampleconsistency/sampleconsistency_test.go]: The whitelisting part is pretty much the same code from the e2e tests, what about moving it to a helper function?

File: test/sampleconsistency/sampleconsistency_test.go (hunk 1 above)
Comment (chaodaiG) [reply test/sampleconsistency/sampleconsistency_test.go]: added

File: test/sampleconsistency/sampleconsistency_test.go (hunk 4 above)
Comment (chaodaiG) [reply test/sampleconsistency/sampleconsistency_test.go]: done

**Comment by chaodaiG** [comment]
> Please elaborate on this test in the initial comment, and add it also as a comment in the code. It's pretty easy to eagerly understand this PR as "this ensures the sample docs are completely in sync with the real code", while actually it ensures that the real code of the samples is properly embedded in the docs (which is great). Even if not yet a full solution for keeping up-to-date documentation, you should make people (especially the docs WG) aware of this, so we can start thinking about going beyond.

Added explanations in both places, please review again @adrcunha

**Pull request merged**